Updates to ChadoStorage -- path argument and cleanup.

spficklin opened the pull request:
<!--- Thank you for contributing! -->
<!--- Provide a general summary of your changes in the Title above -->
<!--- See our Contribution Guidelines here:
          https://tripaldoc.readthedocs.io/en/latest/contributing.html -->


<!---  Please set the header below based on the PR type:
# New Feature
# Bug Fix
# Tripal 4 Core Dev Task  --->

# Tripal 4 Core Dev Task

### Issue #1719

<!--- Enter the Tripal version this PR applies to (i.e. either 3 or 4 ;-p) --->
### Tripal Version: 4.x

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
This PR fixes the changes with joining tables that were described in issue #1719. The fix is needed to make publishing work as we need better specification for joins in order to find records.   The solution proposed on that issue was to rework field properties so that they have a `path` element and a `table_alias_mapping` element that allows developers to fully specify table aliases on joins.  

In the process of adding in this functionality I took the time to fix multiple problems with the way ChadoStorage was handling keep track of how queries were to be handled.  So, this PR includes a major overhaul of the ChadoStorage and introduces a new ChadoRecords class. The ChadoRecords class keeps manages the "records" array that we were using before to keep track of how queries were handled. It provides better error checking, notification of errors, and a better structured array that is more intuitive and useful for debugging.  

@laceysanderson I did have to comment out two functional tests because I think they are buggy. I have makred them with a @todo and my initials (spf) so you can take a look and confirm or correct me.  

## Testing?
<!--- Please describe in detail how to test these changes. -->
<!--- Reviewers will use this section to test the submission! -->
<!--- If you've implemented PHPUnit tests, you can describe the test cases here. -->
To test this PR, you will have to test use of every field. Here is the list of tests you can perform to do so.

1. Create two or more "analysis" content types.  
    1. Tests the `ChadoStorage::insertValues()` for single value fields
    2. Tests the `ChadoStraoge::loadValues()` for single value fields
    3. The following fields are tested:  , `ChadoStringTypeDefault` and `ChadoTextTypeDefault`.
2. Edit the analysis and change all of the values and save.
    1. Tests the `ChadoStorage::updateValues()` for single value fields 
4. Create at least two "organism" content types.
    1.  Repeats the insert and load tests of the analysis above, but you need an organism to create other conten types.
5. Create a two or more "contact" content types. We will use these later for tesing.
6. Create a "gene" and only fill all of the values.  Add as many contact as you want (be sure to add at least two). Add as many synonyms as you like.  Be sure to add at least two.  Be sure to add a sequence because we want to test that the md5checksum and length are calculated.  Be sure to check as least one of the checkboxes at the bottom for the form.
    1.  This tests the following additional fields:  `ChadoBooleanTypeDefault`, `ChadoAdditionalTypeTypeDefault`, `ChadoOrganismTypeDefault`, `ChadoSequenceChecksumTypeDefault`, `ChadoSequenceLengthTypeDefault`, `ChadoSynonymTypeDefault`
    2. Results should show that fields that have joins to another table work (e.g., the organism name).
    3. Results should show that fields with multiple tables in a join work (eg., the synonyms).
    4. Results should show that the md5checksum and sequence length are calculated.
    5. Results should show that fields with multiple cardinality (e.g. contacts and synonyms) store and display all of the values.
7.  Edit the field and change as many values as you like.  Be sure to change the order of the contacts and the synonym.  After saving all of your changes should be visible.
8. Add two new properties to the Gene content type via Admin > Structure > Tripoal Content Types.  Select "Manage Fields" for the "Gene" Content type and click the button to "Create a new field".  Select to create a new "Chado Property".  Create one property with the label of "Note" and another with the label "Comment". Select a controlled vocabulary term for each that is appropiate but different.  Make sure the base table for both is the `feature` table.
9. Return to the gene and edit it.  You should now see to new "Note" and "Comment" fields. Add as many values to both of those.  Be sure to add at least two to each of them. Save the entity.
    1. This tests these additional fields:  `ChadoLinkerPropertyTypeDefault`
    2. Results should show that you can add as many propery values as you like.  You should be able to reorder those while editing as well.
    3. Results should show that you can have multiple properties on the same content type, for the same table  and the do not intefere with one another.
11. Create an "array design" content type.    Save and then edit it at as you like.
      1.  Results should show that you can have two fields both with joins to the same table (e.g. cvterm), via the substrate type and platform type fields, without interferring with one another.
13. Next we will test publishing.  Currently, ChadoStorage supports publishing of two types of content types: those where all fields from a single table are for the same type (e.g. organism), and those where the type is set in a `type_id` column of the base table (e.g. feature, stock).   Go to the Tripal Content List page and delete all of your content types you made above.
14.  Once removed, click the "Publish Tripal Content" button and select "Organism" as the type to publish.  Run the Tripal job. It should restore the two or more organisms that you created previously.
       1. This tests the `ChadoStorage::findValues()` function.
       2. Results should show that you can publish values where the Chado table represents a single content type.
15. Do the same for the gene content type.
      1. Results should show that you can publish values where Chado table distinguishes between contnet  types using the `type_id` column in the base table.  
17.

laceysanderson committed:
Fix some yaml issues.

spficklin committed:
more cleanup and bug fixes

spficklin committed:
Merge branch 'tv4g1-issue1719-usePathForAllActions' of github.com:tripal/tripal into tv4g1-issue1719-usePathForAllActions

spficklin committed:
Fixed bug in adding joins. Removed broken test

spficklin committed:
More bug fixes

spficklin committed:
Fixes properites with cardinality > 1

spficklin committed:
removed dpms, fixed a few tests, commented out one test, and all tests seem to be passing for ChadoStorage. Will run again to be sure.

spficklin committed:
Fixed bug with publishing

spficklin committed:
Fixed merge conflict with 4x.

spficklin committed:
More fixes from botched merge

spficklin commented:
A note:  this PR uses the branch `tv4g1-issue1719-usePathForAllActions-v2` because merging in 4.x on my first attempt was a disaster.  So the `tv4g1-issue1719-usePathForAllActions` is borked and should be ignored going forward.

spficklin commented:
Also, I think we should NOT merge any other PRs into the 4.x branch before this one.  The mishap where the `tv4g1-issue1719-usePathForAllActions` was accidentally merged into 4.x and then files were reverted back to restore it, created challenges with merging 4.x back into this branch. If we don't get this merged before anything else then I'm worried I'll have to repeat fixing the merge conflicts again. It wasn't fun to fix.

laceysanderson committed:
testStorePKeyAction with alias now works.

laceysanderson committed:
Fixed store_link test by setting project_id.

laceysanderson commented:
I've added some updates to the tests that Stephen mentioned having to comment out which fixes them. All the original tests are now being run and passing. Specifically,

- testStorePKeyAction: there is a data provider that tests table alias' for the store_pkey action. There was a mistake in the field definition that got introduced at an unknown point. Essentially, the base_table for the field needs to be different from the table the store_pkey is on (in this case cv vs. db) and no path between the two tables need to exist.
- testStoreLinkAction: there were a couple of bugs introduced into this test at some point that are now fixed. I've made the expected_values array more clear and more complete and set the project record_id for the update.

dsenalik reviewed (approved):
I merged this branch into my dbxref field #1710 and tested including the dbxref field.
It took a while but everything worked as expected, including dbxrefs.
A minor thing, when publishing, if you run two publish jobs at the same time, the second one fails. This is almost certainly the same bug as #1716 so will not affect the approval of this PR.

spficklin commented:
Hey @dsenalik .... just curious what do you mean it "took awhile"...

risharde commented:
Sorry for the delay, I have an issue which may not be related to your branch @spficklin 
I just did a reinstall of Tripal V4 under your branch, I created a chado instance and ran the prepare step
I don't see the Tripal Content Types appearing - could this be because I'm using Drupal 9.x ?
Confirming the branch is up-to-date on my side
Confirming that I ran drush cr to clear cache - did not work
![image](https://github.com/tripal/tripal/assets/33866403/fcb9e7c5-7107-4c9e-984c-64dd83d815d9)

dsenalik commented:
@spficklin just all those testing steps LOL

spficklin commented:
Hey @risharde .  The content type creation was moved out of the prepare step.  To add them, go to Admin > Structure > Tripal Content Types > Click the "import type collection" button.  On the page that appears select all of the content types and click the button. That will add a job.  Unfortunately it won't give you the command-line to run it but you should be able to use the last one you did and just update the job number to the next one.

dsenalik commented:
@risharde if you are not using the docker, you will have to first create the content types by going to "Page Structure" -> "+ Import type collection"

risharde commented:
@spficklin @dsenalik thanks guys, worked like a charm!

risharde commented:
1. Works perfectly! Encountered absolutely no issues - no errors in reports!
2. Works perfectly! No errors in reports as well!
3. Works perfectly! No errors in reports as well!
4. Works perfectly (as expected!) 
5. Works perfectly! (displayed correctly signifying joins worked), md5 checksums generated, length created, checkboxes store as True etc!
6. Works perfectly! (order changes were seen, other changes were seen)
7. Works perfectly!
8. Works great! (Reordering could be seen, edits happened!) Note: Saving was noticeably slower than other task tests so I assume many operations are happening in the background to make this feature work, just a note!
9. Works perfectly! (Creation and editing could be seen!)
10. Worked perfectly! (Items were deleted)
11. Worked perfectly! (This was quite fast! Organisms were added back to the UI!)
12. @spficklin Semi-works (content was recreated but multi-value fields were not fully populated example there was only 1 contact recreated, 1 comment recreated, 1 note recreated even though the previous content added contained 2 contacts, 3 comments and 3 notes.

This was a super exciting experience to see this part of Tripal working! @spficklin I'm not sure if you are able to test this on your side to confirm if #12 is something worth looking into?

dsenalik commented:
You paid better attention than I did @risharde - I confirm that a gene with two contacts, deleted, and published now only shows one contact. 
![20240106_gene_contact](https://github.com/tripal/tripal/assets/8419404/4162960b-05e7-40c0-99f1-cc623ff64c83)

I'll "quick" build a 4.x docker and see if it is the same there...

dsenalik commented:
Results: Under 4.x the contact field is not yet supported.
![20240106_4 x-publish](https://github.com/tripal/tripal/assets/8419404/b37a2121-cb42-48f8-915d-24df9da7798e)

spficklin commented:
Hey @risharde.  I confirm that test number 12 has a problem.  I'll check it.
@dsenalik I can't reproduce that error. How did you generate it?

spficklin committed:
Fixed issue of publishing fields with cardinality > 1

spficklin committed:
Fixed another bug with publishing fields with cardinality > 2 and fixed the Chado Property field so that multiple properties on the same entity get published properly

spficklin commented:
Okay, problems with test number 12 are fixed!  it was actually a deeper problem to fix.  The TripalPublish class didn't support fields with cardinality > 1 so when I asked you all to test it I had forgotten I hadn't fixed that and in my head I thought I had.  So, it's now fixed and that test should work.  

I also had to fix the Chado Property field because it wasn't setting the value for the `type_id` property which caused a problem with publishing.  It now sets it similar to the Additional Type field.   If you want to test it on the same Tripal instance you did your other testing then you have to reset the settings on the Chado Property field.  To do this first clear the cache, then go to Admin > Structure > Tripal Content Types. Select "manage fields" under the Gene contnet type. Then edit the "Note" and "Comment" fields and simply save them (no need to make changes).  This will cause the settings to update.

spficklin commented:
I should be clear.  This PR is ready again for testing test 12 again.

dsenalik commented:
A brand new docker.
Adding the first content, trying to add any of these four Organism, Contact, Protocol, Publication gives this error
`Warning: Undefined array key 1 in Drupal\tripal_chado\Plugin\Field\FieldWidget\ChadoAdditionalTypeWidgetDefault->formElement() (line 66 of modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldWidget/ChadoAdditionalTypeWidgetDefault.php). `

This happens because `$fixed_value` is specified as a boolean TRUE for some fields `$settings['fixed_value'] = TRUE;` which is stored as `1` actually, and for some reason this is getting evaluated in the additional type field, which expects a string like `DB:accession`. In the past, only the additional type field supported this fixed value. It looks like you added this to other fields for publishing. I'm not sure what the best fix is then.
I will try commenting these booleans out and build a new docker to see what happens...
```
    // If this field needs to set a fixed value, set this to TRUE.
    // It indicates to the publishing step to include this field.
    // If not set, then the publishing step may not be able to find matches
    // for this field based on the fixed value.
//    $settings['fixed_value'] = TRUE;
``

laceysanderson commented on tripal/src/TripalStorage/TripalStorageBase.php:
```diff
@@ -207,23 +207,43 @@ public function getStoredValues() {
+    $values[$field_name][$num_items] = [];
+    foreach ($values[$field_name][0] as $key => $value) {
+      $values[$field_name][$num_items][$key] = [];
+      $values[$field_name][$num_items][$key]['value'] = clone $value['value'];
```
This line worries me because if this function is called after the value has been set for delta 0 then it will get cloned into other delta. This could cause bleeding of values between delta.

Perhaps we just set the value to empty as well 🤔 
```suggestion
      $values[$field_name][$num_items][$key]['value'] = clone $value['value'];
      $values[$field_name][$num_items][$key]['value']->setValue(NULL);
```

spficklin replied:
Yeah, I tried that originally but it wouldn't work. I forget why.  My recollection is that each property has checks and wouldn't allow a null in all cases.  I can try again.

spficklin replied:
I was wrong. This fix does work. I do remember having it this way at one point but it broke so I took out the `setValue()`.  But I must have confused myself on something.  It's working with it so I'm accepting.

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoAnalysisTypeDefault.php:
```diff
@@ -147,60 +147,70 @@ public static function tripalTypes($field_definition) {
         'drupal_store' => TRUE,
-        'chado_table' => $linker_table,
-        'chado_column' => $linker_pkey_col,
+        'path' => $linker_table . '.' . $linker_pkey_col,
```
Both the change and the original should work but I think ideally we specify the full join here to ensure that these properties are evaluated as using the same join path. Plus the original doesn't provide any additional information than before `path` was added to all properties.

```suggestion
        'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_left_col . ';' . $linker_pkey_col,
```

spficklin replied:
Yes, my original thought is that we should specify the entire join too but in the end it doesn't matter.  We can't guarantee that someone will use the full path to the base table and since we didn't specify that before and I wanted to make as little changes as possible to the fields so ChadoStorage supports it.  I think if we leave it and we ever break something having a mix will ensure we always support it either way.

spficklin replied:
There is also an advantage in using the shorter path length by not including the base table... as it's more readable.

spficklin replied:
But I'll leave it up to you. I really don't care... I'm just nervous about breaking things and having to spend more time on this PR

laceysanderson replied:
True, I just always think of the fields in core as the "Best Practices" example so want to make sure we give consistent and clear examples 🤔 I lean towards using the full path here but also don't want to break things again 😬 How about we leave these full paths out of this PR. I can make a separate PR that makes them consistent and hopefully that can get in before release but at least by putting them in a separate PR, if something goes wrong it doesn't hold this up.

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoAnalysisTypeDefault.php:
```diff
@@ -147,60 +147,70 @@ public static function tripalTypes($field_definition) {
         'drupal_store' => TRUE,
-        'chado_table' => $linker_table,
-        'chado_column' => $linker_fkey_col,
+        'path' => $linker_table . '.' . $linker_fkey_col,
```
```suggestion
        'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_left_col . ';' $linker_fkey_col,
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoAnalysisTypeDefault.php:
```diff
@@ -147,60 +147,70 @@ public static function tripalTypes($field_definition) {
           'drupal_store' => FALSE,
-          'chado_table' => $linker_table,
-          'chado_column' => $column,
+          'path' => $linker_table . '.' . $column,
```
```suggestion
          'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_left_col . ';' . $column,
```

dsenalik commented:
New docker with the boolean `$fixed_value`s commented out.
 * I also made a change in `ChadoAdditionalTypeWidgetDefault.php` to allow for no default value for `fixed_value`:
```
-    $fixed_value = $field_settings['fixed_value'];
+    $fixed_value = $field_settings['fixed_value'] ?? null;
```

 * Publishing a gene there is an undefined variable, and as a result the total number of field values is not updated at the end.
```
...
 [notice] Step  6 of 6: Add field items to published entities...
  Checking for published items for the field: field_gene_comment...
 [notice]   Checking for published items for the field: field_gene_comment...
  Publishing items 4 for field: field_gene_comment...
 [notice]   Publishing items 4 for field: field_gene_comment...
 [warning] Undefined variable $num_new_items TripalPublish.php:886
  Checking for published items for the field: field_gene_note...
 [notice]   Checking for published items for the field: field_gene_note...
  Publishing items 4 for field: field_gene_note...
 [notice]   Publishing items 4 for field: field_gene_note...
 [warning] Undefined variable $num_new_items TripalPublish.php:886
  Checking for published items for the field: gene_contact...
...
Published 1 new entities, and 0 field values.
 [notice] Published 1 new entities, and 0 field values.
```

 * When I published the gene, the property values "leaked through"
Is this because I removed the fixed_value = true?
Before
![20240107_genebeforepublish](https://github.com/tripal/tripal/assets/8419404/eff56ad4-92eb-467b-964c-ad780e136bfb)

After deletion and publishing
![20240107_geneafterpublish](https://github.com/tripal/tripal/assets/8419404/5f450e5c-c9e0-4b9a-ab3f-f0bb3a048ac9)

 * But all of the other gene fields including both contacts did publish this time 🎉

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoContactTypeDefault.php:
```diff
@@ -164,26 +166,29 @@ public static function tripalTypes($field_definition) {
         'drupal_store' => TRUE,
-        'chado_table' => $linker_table,
-        'chado_column' => $linker_pkey_col,
+        'path' => $linker_table . '.' . $linker_pkey_col,
```
```suggestion
        'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_left_col . ';' . $linker_pkey_col,
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoContactTypeDefault.php:
```diff
@@ -164,26 +166,29 @@ public static function tripalTypes($field_definition) {
         'drupal_store' => TRUE,
-        'chado_table' => $linker_table,
-        'chado_column' => $linker_fkey_col,
+        'path' => $linker_table . '.' . $linker_fkey_col,
```
```suggestion
        'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_left_col . ';' . $linker_fkey_col,
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoContactTypeDefault.php:
```diff
@@ -195,8 +200,9 @@ public static function tripalTypes($field_definition) {
           'drupal_store' => FALSE,
-          'chado_table' => $linker_table,
-          'chado_column' => $column,
+          'path' => $linker_table . '.' . $column,
```
```suggestion
          'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_left_col . ';' . $column,
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoLinkerPropertyTypeDefault.php:
```diff
@@ -80,52 +93,96 @@ public static function tripalTypes($field_definition) {
   }
 
+  /**
+   * We need to set the type_id property value to match the cvter,_id.
```
```suggestion
   * We need to set the type_id property value to match the cvterm_id.
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoSynonymTypeDefault.php:
```diff
@@ -167,47 +168,53 @@ public static function tripalTypes($field_definition) {
       'drupal_store' => TRUE,
-      'chado_table' => $linker_table,
-      'chado_column' => $linker_table_pkey,
+      'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_table_pkey,
```
```suggestion
      'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_fkey_column . ';' . $linker_table_pkey,
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoSynonymTypeDefault.php:
```diff
@@ -167,47 +168,53 @@ public static function tripalTypes($field_definition) {
       'drupal_store' => TRUE,
-      'chado_table' => $linker_table,
-      'chado_column' => 'synonym_id',
+      'path' => $linker_table . '.synonym_id',
```
```suggestion
      'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_fkey_column . ';synonym_id',
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoSynonymTypeDefault.php:
```diff
@@ -167,47 +168,53 @@ public static function tripalTypes($field_definition) {
     $properties[] = new ChadoBoolStoragePropertyType($entity_type_id, self::$id, 'is_current', $linker_is_current_term, [
       'action' => 'store',
-      'chado_table' => $linker_table,
+      'path' => $linker_table . '.is_current',
```
```suggestion
      'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_fkey_column . ';is_current',
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoSynonymTypeDefault.php:
```diff
@@ -167,47 +168,53 @@ public static function tripalTypes($field_definition) {
     $properties[] = new ChadoBoolStoragePropertyType($entity_type_id, self::$id, 'is_internal', $linker_is_internal_term, [
       'action' => 'store',
-      'chado_table' => $linker_table,
+      'path' => $linker_table . '.is_internal',
```
```suggestion
      'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_fkey_column . ';is_internal',
```

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoSynonymTypeDefault.php:
```diff
@@ -167,47 +168,53 @@ public static function tripalTypes($field_definition) {
     // E.g. feature_synonym.pub_id
     $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_pub_id' , $linker_pub_id_term, [
       'action' => 'store',
+      'path' => $linker_table . '.pub_id',
```
```suggestion
      'path' => $base_table . '.' . $base_pkey_col . '>' . $linker_table . '.' . $linker_fkey_column . ';pub_id',
```

laceysanderson commented on tripal_chado/src/Services/ChadoFieldDebugger.php:
```diff
@@ -140,21 +142,18 @@ public function summarizeChadoStorage($chadostorage, $message) {
-   *   A simple array where the key is the chado table name and the value
-   *   is the known value of the primary key for the specified record.
    * @param array $records
    *   This is a nested array describing the records which will need to be
```
```suggestion
   *   This is an instance of the TripalStorage ChadoRecords class which contains
       all the information of records to be inserted/modifiedin chado. 
```

laceysanderson reviewed (commented):
I combed through the code for all but the ChadoStorage and ChadoRecord class. This is super exciting 😍 Here are some things that jumped out at me.

laceysanderson commented:
@dsenalik can you check PR #1730? I think this fixes the discrepancy that you were mentioning for fixed_value. Also @spficklin do you agree that this is the right approach based on how you designed it?

spficklin committed:
Update tripal_chado/src/Plugin/Field/FieldType/ChadoLinkerPropertyTypeDefault.php

Co-authored-by: Lacey-Anne Sanderson <lacey.sanderson@usask.ca>

spficklin committed:
Update tripal_chado/src/Services/ChadoFieldDebugger.php

Co-authored-by: Lacey-Anne Sanderson <lacey.sanderson@usask.ca>

spficklin commented:
Thanks @laceysanderson for your review. I committed some and added comments for you to review.

@dsenalik . I'm baffeled why you're getting those errors.  The code you have I"m not sure matches what I'm committing for some reason. I recognize those errors you're getting but they are things I've already fixed and I can't reproduce them in my code.  I can't tell if Lacey's comment just before this means she resolved the issue or not.  But if not can you check to make sure you have the most up-to-date code in the same branch?

spficklin committed:
Update tripal/src/TripalStorage/TripalStorageBase.php

Co-authored-by: Lacey-Anne Sanderson <lacey.sanderson@usask.ca>

laceysanderson commented:
> @dsenalik . I'm baffeled why you're getting those errors. The code you have I"m not sure matches what I'm committing for some reason. I recognize those errors you're getting but they are things I've already fixed and I can't reproduce them in my code. I can't tell if Lacey's comment just before this means she resolved the issue or not. But if not can you check to make sure you have the most up-to-date code in the same branch?

I did see the same errors that @dsenalik did on the most recent commit from this branch and a fresh docker. That's why I tried to solve them in the other PR.

Specifically, I also saw the following errros when adding any of these four Organism, Contact, Protocol, Publication:

> Warning: Undefined array key 1 in Drupal\tripal_chado\Plugin\Field\FieldWidget\ChadoAdditionalTypeWidgetDefault->formElement() (line 66 of modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldWidget/ChadoAdditionalTypeWidgetDefault.php).

spficklin commented:
> Specifically, I also saw the following errros when adding any of these four Organism, Contact, Protocol, Publication:

You see these when manually adding via the GUI?  Do you get a warning on the screen or this appears in the recent logs?

dsenalik commented:
You see this error directly on screen after trying to add these types via the GUI

spficklin commented:
I can confirm that I do not get these errors.....

dsenalik commented:
I wonder if it is a PHP version thing then? I always test on the latest available, 8.2
Edit: no, this code throws the error both in php 8.1 and 8.2
```
$fixed_value = "1";
list($idSpace, $accession) = explode(':', $fixed_value);
```

spficklin committed:
Fix for warning message about missing key 1

spficklin commented:
Hi @dsenalik I think I know how to fix it even thought I don't see the warning.  I'm on PHP 8.1 and  I usually get warnings about missing keys so I'm not sure why I'm not getting it in this case.  I just pushed a fix that I think will work. Can you try it?

dsenalik commented:
Well commit https://github.com/tripal/tripal/pull/1729/commits/99cfebb595b28b4f7accb1118136e336d8612c1c has fixed this error, I do not see it any more on a fresh docker.

spficklin commented:
I think this PR is ready for merge after @laceysanderson responds to the comments above.

risharde commented:
I just pulled in the latest changes, drush cr, created a new gene, deleted it, ran tripal publish and it didn't work for me - none of the fields appeared except Title
![image](https://github.com/tripal/tripal/assets/33866403/5d43239c-0851-40b0-b234-38865453c540)

I'm going to do a full reinstall and test back

dsenalik committed:
fix variable name num_new_items -> num_field_items in publish

dsenalik commented:
In the interests of time I just pushed a single typo fix, hope it doesn't confuse anything https://github.com/tripal/tripal/pull/1729/commits/22d612e5921559b28735450df485b26962e00335

This removes the undefined variable when publishing from comment https://github.com/tripal/tripal/pull/1729#issuecomment-1880093532
Value is now updated: `Published 1 new entities, and 13 field values.`

dsenalik commented:
I just tested on my most recent docker, and created an organism, two contacts, and then a gene.
Publishing that appeared to work fine.

Then I added the two properties, Note, and Comment from step 8, and added two of each on a second new gene.
When I went to publish this one - woe is us!
```
...
2024-01-07 17:01:27
Tripal Job Launcher
Running as user 'drupaladmin'
-------------------
2024-01-07 17:01:27: Job ID 3.
2024-01-07 17:01:27: Calling: tripal_publish(gene, chado_storage, Array)
Step  1 of 6: Find matching records... 
 [notice] Step  1 of 6: Find matching records... 
Step  2 of 6: Generate page titles...
 [notice] Step  2 of 6: Generate page titles...
Step  3 of 6: Find existing published entities...
 [notice] Step  3 of 6: Find existing published entities...
Step  4 of 6: Publishing 1 new entities...
 [notice] Step  4 of 6: Publishing 1 new entities...
Step  5 of 6: Find IDs of entities...
 [notice] Step  5 of 6: Find IDs of entities...
Step  6 of 6: Add field items to published entities...
 [notice] Step  6 of 6: Add field items to published entities...
  Checking for published items for the field: field_gene_comment...
 [notice]   Checking for published items for the field: field_gene_comment...
  Publishing items 2 for field: field_gene_comment...
 [notice]   Publishing items 2 for field: field_gene_comment...
  Checking for published items for the field: field_gene_note...
 [notice]   Checking for published items for the field: field_gene_note...
  Publishing items 2 for field: field_gene_note...
 [notice]   Publishing items 2 for field: field_gene_note...
  Checking for published items for the field: gene_contact...
 [notice]   Checking for published items for the field: gene_contact...
  Publishing items 4 for field: gene_contact...
 [notice]   Publishing items 4 for field: gene_contact...
 [warning] Undefined array key 2 TripalPublish.php:740
 [warning] Undefined array key "" TripalPublish.php:741
ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "entity_id" of relation "tripal_entity__gene_contact" violates not-null constraint
DETAIL:  Failing row contains (gene, 0, null, 1, und, 0, 2, 3, 2, 2, null, null, null).: 
      INSERT INTO tripal_entity__gene_contact
        (bundle, deleted, entity_id, revision_id, langcode, delta, gene_contact_record_id, gene_contact_linker_id, gene_contact_link, gene_contact_contact_id) VALUES
(:bundle_3, :deleted_3, :entity_id_3, :revision_id_3, :langcode_3, :delta_3, :gene_contact_record_id_3, :gene_contact_linker_id_3, :gene_contact_link_3, :gene_contact_contact_id_3),
(:bundle_4, :deleted_4, :entity_id_4, :revision_id_4, :langcode_4, :delta_4, :gene_contact_record_id_4, :gene_contact_linker_id_4, :gene_contact_link_4, :gene_contact_contact_id_4); Array
(
    [:bundle_3] => gene
    [:deleted_3] => 0
    [:entity_id_3] => 
    [:revision_id_3] => 1
    [:langcode_3] => und
    [:delta_3] => 0
    [:gene_contact_record_id_3] => 2
    [:gene_contact_linker_id_3] => 3
    [:gene_contact_link_3] => 2
    [:gene_contact_contact_id_3] => 2
    [:bundle_4] => gene
    [:deleted_4] => 0
    [:entity_id_4] => 
    [:revision_id_4] => 1
    [:langcode_4] => und
    [:delta_4] => 1
    [:gene_contact_record_id_4] => 2
    [:gene_contact_linker_id_4] => 4
    [:gene_contact_link_4] => 2
    [:gene_contact_contact_id_4] => 3
)

 [error]  ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "entity_id" of relation "tripal_entity__gene_contact" violates not-null constraint
DETAIL:  Failing row contains (gene, 0, null, 1, und, 0, 2, 3, 2, 2, null, null, null).: 
      INSERT INTO tripal_entity__gene_contact
        (bundle, deleted, entity_id, revision_id, langcode, delta, gene_contact_record_id, gene_contact_linker_id, gene_contact_link, gene_contact_contact_id) VALUES
(:bundle_3, :deleted_3, :entity_id_3, :revision_id_3, :langcode_3, :delta_3, :gene_contact_record_id_3, :gene_contact_linker_id_3, :gene_contact_link_3, :gene_contact_contact_id_3),
(:bundle_4, :deleted_4, :entity_id_4, :revision_id_4, :langcode_4, :delta_4, :gene_contact_record_id_4, :gene_contact_linker_id_4, :gene_contact_link_4, :gene_contact_contact_id_4); Array
(
    [:bundle_3] => gene
    [:deleted_3] => 0
    [:entity_id_3] => 
    [:revision_id_3] => 1
    [:langcode_3] => und
    [:delta_3] => 0
    [:gene_contact_record_id_3] => 2
    [:gene_contact_linker_id_3] => 3
    [:gene_contact_link_3] => 2
    [:gene_contact_contact_id_3] => 2
    [:bundle_4] => gene
    [:deleted_4] => 0
    [:entity_id_4] => 
    [:revision_id_4] => 1
    [:langcode_4] => und
    [:delta_4] => 1
    [:gene_contact_record_id_4] => 2
    [:gene_contact_linker_id_4] => 4
    [:gene_contact_link_4] => 2
    [:gene_contact_contact_id_4] => 3
)
 
[site http://default] [TRIPAL] ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "entity_id" of relation "tripal_entity__gene_contact" violates not-null constraintDETAIL:  Failing row contains (gene, 0, null, 1, und, 0, 2, 3, 2, 2, null, null, null).:       INSERT INTO tripal_entity__gene_contact        (bundle, deleted, entity_id, revision_id, langcode, delta, gene_contact_record_id, gene_contact_linker_id, gene_contact_link, gene_contact_contact_id) VALUES(:bundle_3, :deleted_3, :entity_id_3, :revision_id_3, :langcode_3, :delta_3, :gene_contact_record_id_3, :gene_contact_linker_id_3, :gene_contact_link_3, :gene_contact_contact_id_3),(:bundle_4, :deleted_4, :entity_id_4, :revision_id_4, :langcode_4, :delta_4, :gene_contact_record_id_4, :gene_contact_linker_id_4, :gene_contact_link_4, :gene_contact_contact_id_4); Array(    [:bundle_3] => gene    [:deleted_3] => 0    [:entity_id_3] =>     [:revision_id_3] => 1    [:langcode_3] => und    [:delta_3] => 0    [:gene_contact_record_id_3] => 2    [:gene_contact_linker_id_3] => 3    [:gene_contact_link_3] => 2    [:gene_contact_contact_id_3] => 2    [:bundle_4] => gene    [:deleted_4] => 0    [:entity_id_4] =>     [:revision_id_4] => 1    [:langcode_4] => und    [:delta_4] => 1    [:gene_contact_record_id_4] => 2    [:gene_contact_linker_id_4] => 4    [:gene_contact_link_4] => 2    [:gene_contact_contact_id_4] => 3)
```

This is all that got published
![20240107_second_gene](https://github.com/tripal/tripal/assets/8419404/7ca5b6b0-dc15-4dbb-a3f1-1ef8b0ca829d)

risharde commented:
@dsenalik pulled your last commit, reinstalled, created 2 contacts, 1 organism, created gene, deleted it and published it back - noticing this

![image](https://github.com/tripal/tripal/assets/33866403/92084c73-4129-4e32-843c-255607d47351)

so confirming it doesn't work on my side either

Edit:
Notes should have been 3 entries with notes 1, notes 2, notes 3
Comments should have been 3 entries with comments 1, comments 2, comments 3

dsenalik commented:
@risharde but you didn't get any errors when publishing 😕 ?

risharde commented:
@dsenalik No errors there, it created the Gene page
One other thing I notice is that the author is not attached - not sure if this is super important but just mentioning it
![image](https://github.com/tripal/tripal/assets/33866403/57cf0b48-e8db-43ba-b3b2-88e9c1a905cd)

dsenalik commented:
I noticed that too, but I assume it is expected behavior, as a published page does not have an "author".

spficklin commented:
I just tested it on my end with all of the recent fixes and it seems to be working just fine :-/  I'll try a complete reinstall and see if that makes a difference.

dsenalik commented:
If it helps, this part of the publish log
```
 [notice]   Publishing items 4 for field: gene_contact...
 [warning] Undefined array key 2 TripalPublish.php:740
 [warning] Undefined array key "" TripalPublish.php:741
```
There should only have been **2** items for gene_contact, not sure why it tried to do **4**. That's how many properties there were though...

spficklin commented:
But you all didn't have these problems prior to our few tweaks, right?

dsenalik commented:
As I recall, and looking at previous comments https://github.com/tripal/tripal/pull/1729#pullrequestreview-1807523715, no this did not happen earlier.

I just now tried a new gene with three each of note and comment, and as I suspected, somehow 3+3 showed up under gene_contact:
```
 [notice]   Publishing items 6 for field: gene_contact...
 [warning] Undefined array key 4 TripalPublish.php:740
 [warning] Undefined array key "" TripalPublish.php:741
```
![20240107_threeeach](https://github.com/tripal/tripal/assets/8419404/65a3f73d-3f85-4194-a9bc-0a16dfbba844)

spficklin commented:
I'm now seeing the same as @risharde .  Digging into it..... I really want to cry right now.

dsenalik commented:
I bet it is a simple fix 😟 but it might take hours to find...

spficklin committed:
Fixed another issue with the  setting

spficklin commented:
Okay. I think I have it fixed. The problem was related to that pesky `fixed_value`.  We definintely need to redesign that as per the discussinon on PR #1730 .  But for now, I think I have it fixed so things will work.  

I would suggested starting from a fresh install.

risharde commented:
@spficklin It works!!! The order of the multiple values changed but I suspect this is normal behaviour. Great going!! 

![image](https://github.com/tripal/tripal/assets/33866403/559601f4-5bee-4fab-8e48-6b77d4d6c591)

dsenalik commented:
Just tested a new docker, gene content type, three different properties and three values for each property.
Publishing worked perfectly 😀
Also three contacts and two synonyms were published correctly.
@risharde ~~as far as order changing, I don't know if publish yet supports using chado ranks, such as from the `featureprop` table?~~
edit: I just did the experiment, changed the rank in `featureprop` before publishing, and the order was **correct** based on those ranks.
```
sitedb=> update featureprop set rank=-1 where value like '%3';
sitedb=> select * from featureprop order by featureprop_id;
 featureprop_id | feature_id | type_id |   value   | rank 
----------------+------------+---------+-----------+------
             15 |          1 |    3029 | Comment 1 |    0
             16 |          1 |    3029 | Comment 2 |    1
             17 |          1 |    3029 | Comment 3 |   -1
             18 |          1 |     727 | DNA1      |    0
             19 |          1 |     727 | DNA2      |    1
             20 |          1 |     727 | DNA3      |   -1
             21 |          1 |    3143 | Note 1    |    0
             22 |          1 |    3143 | Note 2    |    1
             23 |          1 |    3143 | Note 3    |   -1
(9 rows)
```
![20240108_rank](https://github.com/tripal/tripal/assets/8419404/5ce791ef-03d7-4218-b5ce-6c4d44e78443)

Confirmed a second time with `update featureprop set rank=5 where value like '%1';` and published in 3,2,1 order as expected.

laceysanderson reviewed (approved):
I did one more test with the most recent commit.

I created multiple pieces of content of almost every time, edited them, deleted them and tried to trigger validation errors. Page creation seems very stable 😌 

With 3 reviewers saying this is working and a code review, this can be merged! 🎉 Thanks for all the work @spficklin and thanks for the weekend reviews @risharde and @dsenalik!

One big step closer to release!
